### ISSUE TITLE
[css-scrollbars] Suggestion: "Overflow styling" / Scrollbar styling

### ISSUE THREAD

**Issue by KingSora** [w3c/csswg-drafts#107]
Hello!

I don't really know if this is the correct place to post suggestions, but I will do it anyway.
If this isn't the correct place, please correct me and tell me where I have to post my suggestion.

Ok, so yeah, I've been working on my web-app and there are a few things I don't like about the scrollbars. My biggest problem is, that I have no possibility to style my scrollbars properly. In Chrome / webkit are a few options, but they are prefixed and not supported in all browsers.
As well as the "Overflow: overlay" spec.

So my suggestions are the following:

Webkit did already a good job on styling scrollbars with pseudo elements, so why don't accept them as a standard:

``` css
selector:-webkit-scrollbar
selector:-webkit-scrollbar-button
selector:-webkit-scrollbar-track
selector:-webkit-scrollbar-track-piece
selector:-webkit-scrollbar-thumb
selector:-webkit-scrollbar-corner
selector:-webkit-resizer
```

I like the idea of the different states too

``` css
:horizontal
:vertical
:decrement
:increment
:start
:end 
:double-button
:single-button
:no-button
:corner-present
:window-inactive
```

If you don't know the styling possibilities in webkit, please check this article: https://css-tricks.com/custom-scrollbars-in-webkit/

There is something I really don't like, its the "overflow: overlay" property.
If you don't know the "overflow: overlay" property, follow this link: http://quirksmode.org/css/css2/overflow.html

Basically the overflow property is there for deciding if the element should have a scrollbar or not.
These are the standard values which all browsers support:

``` css
visible - content is not clipped, thats why there cant be any scrollbars
hidden - content is clipped but without scrollbars
scroll - content is clipped but with scrollbars (always visible)
auto - the content is clipped but with scrollbars (visible if content is overflowing)
```

And now the overlay value:

``` css
overlay - content is clipped, scrollbars are always visible but are rendered above the container
```

This kind of value isnt ment to be in the overflow property, it should be a styling option for the ":-webkit-scrollbar" pseudo element:

``` css
selector:-webkit-scrollbar {
position: absolute;
}
```

This would be equal to the "overflow: overlay" property.
Unfortunately the ":-webkit-scrollbar" pseudo element currently doesn't accept the "position" property.

**In short:**
The idea is to implement scrollbars as pseudo elements on elements which are displaying scrollbars by the correct setting of the property "overflow". On the pseudo element ":-webkit-scrollbar" the property "position" can be applied for custom positioning and achieving of the "overflow: overlay" effect. In the final standard the pseudo elements should work without any prefix of course.

I hope you understand my idea of scrollbar styling and the, in my opinion obsolete property "overflow: overlay". Any feedback to my suggestion is highly appreciated! :)

Best regards!

**Comment by tabatkins** [comment]
Yup, adopting WK/Chrome's scrollbar styling has been suggested in the past, but the other browsers haven't been very keen on it so far.  Without anyone else wanting to implement, there's not a lot of reason to spec it.

(That said, I use the properties myself, to get less-intrusive scrollbars on some personal projects.)

> This kind of value isnt ment to be in the overflow property, it should be a styling option for the ":-webkit-scrollbar" pseudo element:

You're right, it doesn't belong in `overflow` itself.  We have a proposal for `overflow-style` which controls more aspects of how the content overflows, but it's been slow to be adopted.  "overlay" hasn't even been accepted by the group yet as a reasonable value to spec.  ¯\_(ツ)_/¯

**Comment by KingSora** [comment]
Alright, thanks for your answer!
In my opinion the `overflow-style` shouldnt be implemented, because the efficiency wouldnt be as good as a scrollbar-style. With my "position: absolute;" suggestion you would have a overlay-scrollbar plus you would be free to position it as you want (example: `top: 0; left : 0; bottom: auto; right: auto;` would position the scrollbar to the left side.

I hope in future the group will realize the importance of scrollbar-styling.

**Comment by frivoal** [comment]
different browser vendors may have a different opinion, but I don't think your suggestion of complete freedom of styling the scroll-bar would get everyone's favor. Browser vendors need not only to give power to authors, but also to make sure things make sense for users. Allowing authors to tweak native controls like scrollbars is considered risky, as the resulting UI may end up making no sense. Not that authors make page that make no sense on purpose, but it isn't rare that they fail to take into account some variations between browsers or platforms, and end up making something that doesn't quite work everywhere.

`overflow-style` would give a more limited set of options. This is certainly more limiting for authors, but it also more likely to always work. I think it is a decent compromise.

**Comment by frivoal** [comment]
Agenda+, to see if we can close this as wontfix for the reasons above.

**Comment by upsuper** [comment]
I'm feeling disappointed on that a feature mistakenly exposed to the web ends up needing to be implemented everywhere for interop.

I don't think doing the scrollbar pseudo-element is a good idea. OSs continuously evolve their scrollbars to provide better user experience. macOS and Ubuntu, IIRC, both have quite different scrollbar structure than that on Windows. Having something like this specified may make it harder for browsers to have decent behavior on different platforms.

Also, different OSs having different scrollbar structure also means testing interop is harder, because you would need to take not only engines but also platforms into account.

Actually, it is already not fully interop among platforms for Blink. `::-webkit-scrollbar-button` doesn't have any effect on macOS while it works on Windows.

>  <dael> TabAtkins: Oh no. We can't remove. They're used more than possible to kill.

I doubt this.

From the document [Blink principles of web compatibility](https://docs.google.com/document/d/1RC-pBBvsazYfCNNUSkPqAVpSpNJ96U8trhNkfV0v9fk/edit#), it seems there are various aspects taken into account when deciding whether Blink wants to commit some breakage, not just the use counter.

It seems to me that removing this
* would mostly be a cosmetic breakage, which is generally safer than other kinds of breakage, and
* may actually give users more consistent experience on their platform, which could improve user experience, and
* aligns browsers behavior with less overall effort.

For web developers, I don't think the current situation, that it is not even fully interop among platforms even within a single engine, is very helpful. Existing usages may already fail to take different platforms into consider.

If we really want this kind of feature, I think we should think about some better design, rather than just accept whatever happens to exist, especially given that its origin thinks it's a bad idea and wants to deprecate it.

**Comment by KingSora** [comment]
@upsuper
I think its important to give the developer a possibility to style his scrollbars, if he wants uniform scrollbars on all systems. 

The final spec could include a scrollbar structure where all elements which a scrollbar must have are listed. That means, if the developer decides to use a styled scrollbar with buttons and scrollbar buttons are in the spec, he has a possibility to style a scrollbar with buttons, regardless wheter the native scrollbar has buttons or not. This would be a "all or nothing" scenario, use a completely styled scrollbar or use the native system scrollbar.

Since the developer would have the full control over styled scrollbars and lets say he doesn't want scrollbar buttons although they are in thr spec, he could set them to "display: none" to "remove" them.

I think webkit did a nice job on their spec structure because you really have access to every important element in every state. But I dont like how they are handling the actual styling. They only allow specific properties which aren't documented. There is definitly much room for improvement.

**Comment by frivoal** [comment]
I agree with @upsuper.  @tabatkins @RByers, could you look more closely into the possibility of unshipping this, not only looking at usage (which is indeed somewaht high), but also including the rest of your more nuanced criteria?

**Comment by tabatkins** [comment]
From what I understand, removing this ends up with far more than a cosmetic breakage, as a number of sites (Facebook, several Google properties) use it with the expectation that they can get thin scrollbars, and reverting them to standard-thickness would end up drastically messing up the UI.

**Comment by frivoal** [comment]
Since this is currently not supported in other UAs, do these sites do UA sniffing to  serve a different UI to browsers lacking that capability?

If not, then there's no problem :)
Either way, we should seriously consider giving authors tools to do feature detection rather than UA sniffing here. Time for `@supports selector(:-webkit-scrollbar) {...}` ?

**Comment by RByers** [comment]
@bokand @dtapuska does the chromium input team own scrollbar styling?  Our principle in general for this sort of thing is is that we should either be moving towards standardization or removal (which in this case probably means better understanding the compat constraints for what behavior is actually required by the web beyond just a few popular origins).

**Comment by bokand** [comment]
No, input generally deals with scrollbars as they relate to handling input events and scrollability issues. For issues with scrollbars appearance that would likely fall to some combination of paint team and layout team. @skobes and @progers

**Comment by skobes** [comment]
I would support removing Chrome's current implementation of custom scrollbars, if possible.

It might be good to have a simpler / less general API, with a path to standardization, that still allows some degree of visual customization of scrollbars, which we could push as an alternative.

It looks like Facebook has cross-browser "custom" scrollbars implemented entirely in DOM/JS, so I don't think they would be affected.

**Comment by tantek** [comment]
In parallel I have been pursuing a minimal spec scoped for compat with existing scrollbar styling properties, starting with just colors, to document existing scrollbar coloring support just enough to gain additional interop implementations and quickly get something dependable for this use-case to RE (separate from everything else we are evolving / iterating in css-ui-4).

The CSSWG discussed this at our f2f today and approved adopting the separate CSS Scrollbars spec that I've written up accordingly as an editor's draft - https://github.com/w3c/csswg-drafts/issues/1955

Given the intended conservative compat scope of CSS Scrollbars, if that's sufficient for this issue, feel free to close.  Otherwise if there's desire / request for use-cases beyond scrollbar colors (and perhaps widths), feel free to document those additional future use-cases here and keep this issue open.

Thanks! -t

**Comment by zocky** [comment]
I'd like to chime in with an obvious use case for overlaid scrollbars which don't reserve gutter space.

Consider a vertically split layout in which you have a scrollable selectable list (or any other content with varying backgrounds) on the left side. A constantly visible vertical scrollbar in the middle of the screen is *very* intrusive in such a layout - you want it to show only while scrolling and on mouse hover.  At the same time, you don't want it to have a gutter showing between your list and the content on the right when the scrollbar is hidden.

A good example is the explorer sidebar in VS Code. Its scrollbar fades in and out on mouse-in mouse-out, but it doesn't take up any space, and the selected list item takes up the full width of the list. It's a good and useful design which functions very well, but it's not achievable with pure CSS.

**Comment by ExE-Boss** [comment]
> `scrollbar‑3dlight‑color`, `scrollbar‑arrow‑color`, `scrollbar‑base‑color`, `scrollbar‑darkshadow‑color`, `scrollbar‑face‑color`, `scrollbar‑highlight‑color`, `scrollbar‑track‑color`, `scrollbar‑shadow‑color` 🤦🏻‍♂️

Yeah, Windows ME scrollbars <i title="sarcasm">obviously won’t</i> look out of place in modern operating systems.<sup>&lt;/sarcasm&gt;</sup>

<img src="https://user-images.githubusercontent.com/3889017/38644029-7fa81098-3ddf-11e8-900a-e5e8a9fedd36.png" alt="Microsoft Edge with Windows ME scrollbars" title="The horror, the horror!" style="max-width:100%;"><br/>The horror, _the horror!_

**Comment by tantek** [comment]
The CSS Scrollbars first public working draft ([https://www.w3.org/TR/2018/WD-css-scrollbars-1-20180925/](https://www.w3.org/TR/2018/WD-css-scrollbars-1-20180925/)) was published, with as requested removal of “scrollbar‑3dlight‑color, scrollbar‑arrow‑color, scrollbar‑base‑color, scrollbar‑darkshadow‑color, scrollbar‑face‑color, scrollbar‑highlight‑color, scrollbar‑track‑color, scrollbar‑shadow‑color”.

Rather than a new 'overflow-style' property, see the CSS Overflow Level 4 ([https://drafts.csswg.org/css-overflow-4/](https://drafts.csswg.org/css-overflow-4/)) specification for how it handles the related use-cases. I believe all consensus use-cases requested in this issue have been addressed, and this issue can be closed without any further edits.

(Originally published at: https://tantek.com/2020/023/t1/)
